Update golangci t0 1.61.0

praveenkumar opened the pull request.

dependabot[bot] committed:
build(deps): bump github.com/golangci/golangci-lint in /tools

Bumps [github.com/golangci/golangci-lint](https://github.com/golangci/golangci-lint) from 1.59.1 to 1.61.0.
- [Release notes](https://github.com/golangci/golangci-lint/releases)
- [Changelog](https://github.com/golangci/golangci-lint/blob/master/CHANGELOG.md)
- [Commits](https://github.com/golangci/golangci-lint/compare/v1.59.1...v1.61.0)

---
updated-dependencies:
- dependency-name: github.com/golangci/golangci-lint
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

praveenkumar committed:
tools/go.mod: Update the go version from 1.22.0 to 1.22.1

Latest release of golangci-lint uses `go 1.22.1` as part of go mod file
so when our tooling try to consume this version then following is added
for `tools/go.mod` because in our mod file we mention minimum required
go version is `1.22.0` and as per go toolchain guideline it should run
in in newer toolchain

```
go 1.22.0
toolchain go1.22.5
```

snip from: https://go.dev/doc/toolchain
```
When the go or toolchain line is newer than the bundled toolchain,
the go command runs the newer toolchain instead. For example, when
using the go command bundled with Go 1.21.3 in a main module that
says go 1.21.9, the go command finds and runs Go 1.21.9 instead. It
first looks in the PATH for a program named go1.21.9 and otherwise
downloads and caches a copy of the Go 1.21.9 toolchain.
```

- https://github.com/golangci/golangci-lint/blob/v1.61.0/go.mod#L3

cfergeau commented on update-go-version.sh:
```diff
@@ -10,7 +10,7 @@ golang_base_version=$1
 
 go mod edit -go "${golang_base_version}.0"
-go mod edit -go "${golang_base_version}.0" tools/go.mod
+go mod edit -go "${golang_base_version}.1" tools/go.mod
```
~~better to use `"${golang_base_version}.*"` or `"${golang_base_version}.[0-9]+"` so that we don't keep updating it for each minor release.~~

I should read the full command, not just its arg :), ignore this

cfergeau reviewed (commented):
This introduces new linting failures.

praveenkumar committed:
shell32_windows.go: use int32 instead of int for showCmd parameter

Looks like it is just the to pass value if command is shown or not and
`int32` can handle it. Also it will fix the following golanglint error.

```
pkg/os/windows/win32/shell32_windows.go:40:77: G115: integer overflow conversion int -> int32 (gosec)
	return windows.ShellExecute(hwnd, op, toUint16ptr(file), params, dir, int32(showCmd))

```

praveenkumar committed:
driver_darwin: Make readPidFromFile to return int32 instead int

With that we don't need to do type conversion and avoid following
golint error.

```
pkg/drivers/vfkit/driver_darwin.go:407:40: G115: integer overflow conversion int -> int32 (gosec)
	exists, err := process.PidExists(int32(pid))
	                                      ^
pkg/drivers/vfkit/driver_darwin.go:414:36: G115: integer overflow conversion int -> int32 (gosec)
	p, err := process.NewProcess(int32(pid))
	                                  ^
```

cfergeau commented on pkg/os/shell/shell_windows.go:
```diff
@@ -30,6 +30,7 @@ func getProcessEntry(pid int) (pe *syscall.ProcessEntry32, err error) {
 	}
 
 	for {
+		//#nosec G115 -- pid is coming from os.Getppid() as int
```
the `os.Getppid()` call is 2 level above `getProcessEntry` in the call chain. Nothing guarantees that in the future we are not going to reuse this function with a `pid` value coming from something different.
I'd prefer to change the `getProcessEntry(pid int)` to `getProcessEntry(pid uint32)`, and to change `detect` to convert the value returned by `os.Getppid()` to `uint32`

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -265,8 +265,8 @@ func configureSharedDirs(vm *virtualMachine, sshRunner *crcssh.Runner) error {
 	telemetry.SetCPUs(ctx, startConfig.CPUs)
-	telemetry.SetMemory(ctx, uint64(startConfig.Memory)*1024*1024)
-	telemetry.SetDiskSize(ctx, uint64(startConfig.DiskSize)*1024*1024*1024)
+	telemetry.SetMemory(ctx, uint64(startConfig.Memory)*1024*1024)          //#nosec G115 -- validation make sure memory value is not negative
```
actually I haven't been able to find such >=0 checks in the  validation code?

praveenkumar replied:
for memory as part of validation we use the base memory as default preset one (in case of OCP it is 10752 and for microshift 4096

```go
// ValidateMemory checks if provided Memory count is valid
func ValidateMemory(value int, preset crcpreset.Preset) error {
	if value < constants.GetDefaultMemory(preset) {
		return fmt.Errorf("requires memory in MiB >= %d", constants.GetDefaultMemory(preset))
	}
	return ValidateEnoughMemory(value)
```

cfergeau commented on pkg/drivers/hyperv/hyperv_windows.go:
```diff
@@ -74,6 +74,7 @@ func (d *Driver) UpdateConfigRaw(rawConfig []byte) error {
 		}
 	}
 	if newDriver.DiskCapacity != d.DiskCapacity {
+		//#nosec G115 -- validation make sure disk value is in range
```
I'd prefer if code in `pkg/drivers` did not make assumptions regarding how it's used, and in particular that we don't rely on code in `pkg/crc` having to be used before passing data to the drivers.
Moreover, I'm not sure we have the needed range check in validation

praveenkumar committed:
shell_windows: Get pid as uint32 instead int

os.Getppid() return int value but syscall.ProcessEntry32 struct have
`ProcessID` as uint32 so we are doing the type conversion as part of
`os.Getppid()` instead later in stage. It will fix following golint
error.
```
pkg/os/shell/shell_windows.go:33:38: G115: integer overflow conversion int -> uint32 (gosec)
                    if processEntry.ProcessID == uint32(pid) {
```

praveenkumar committed:
libhvee_windows: Use uint64 as arg for disk resize function

It will reduce following error for golint ci.
```
pkg/drivers/libhvee/libhvee_windows.go:92:31: G115: integer overflow conversion uint64 -> int64 (gosec)
		if err := d.resizeDisk(int64(update.DiskCapacity)); err != nil {
		                            ^
pkg/drivers/libhvee/libhvee_windows.go:196:27: G115: integer overflow conversion uint64 -> int64 (gosec)
	return d.resizeDisk(int64(d.DiskCapacity))
	                         ^
pkg/drivers/libhvee/libhvee_windows.go:288:26: G115: integer overflow conversion int64 -> uint64 (gosec)
	newSize := strongunits.B(newSizeBytes)
```

praveenkumar committed:
driver_darwin: Change arg for resize from int64 to uint64

It will fix following golint ci error
```
pkg/drivers/vfkit/driver_darwin.go:135:23: G115: integer overflow conversion uint64 -> int64 (gosec)
	return d.resize(int64(d.DiskCapacity))
	                     ^
pkg/drivers/vfkit/driver_darwin.go:176:7: G115: integer overflow conversion int -> uint (gosec)
		uint(d.CPU),
		    ^
pkg/drivers/vfkit/driver_darwin.go:177:9: G115: integer overflow conversion int -> uint64 (gosec)
		uint64(d.Memory),
		      ^
```

praveenkumar committed:
Update crc-org/machine module

With this update type of memory and cpu is changed to uint from int
which will allow us to fix some of interter overflow issue with latest
golint ci.

praveenkumar committed:
Change type of DiskSize to int to uint

It help us to fix integer overflow issue with latest version of
golint-ci.

praveenkumar committed:
kubeconfig: For string conversion use FormatUint instead Itoa

This will fix follow golint ci warning
```
pkg/crc/machine/kubeconfig.go:189:29: G115: integer overflow conversion uint -> int (gosec)
					port = strconv.Itoa(int(ingressHTTPSPort))
```

praveenkumar commented:
@cfergeau Can you take another look to this PR, it have machine update and also golint ci update.

cfergeau commented on pkg/crc/config/config.go:
```diff
@@ -207,6 +212,13 @@ func (c *Config) Get(key string) SettingValue {
+			return SettingValue{
+				Invalid: true,
+			}
+		}
```
Unrelated to this PR, but wondering if we could use go generics to have a generic 'cast' function instead of switching over the types.

praveenkumar replied:
Right, that would be a good exercise, we can put a tracker issue for it.

cfergeau commented on cmd/crc/cmd/start.go:
```diff
@@ -39,7 +39,7 @@ func init() {
 	flagSet.UintP(crcConfig.CPUs, "c", constants.GetDefaultCPUs(crcConfig.GetPreset(config)), "Number of CPU cores to allocate to the instance")
 	flagSet.UintP(crcConfig.Memory, "m", constants.GetDefaultMemory(crcConfig.GetPreset(config)), "MiB of memory to allocate to the instance")
-	flagSet.UintP(crcConfig.DiskSize, "d", constants.DefaultDiskSize, "Total size in GiB of the disk used by the instance")
+	flagSet.Uint64P(crcConfig.DiskSize, "d", constants.DefaultDiskSize, "Total size in GiB of the disk used by the instance")
```
I'd prefer to keep DiskSize as uint as long as it's holding GiBs. When we convert it to bytes, then it makes sense to use an uint64. I don't understand how all the changes in this commit help to fix unspecified "integer overflow issue".

praveenkumar replied:
Done, switched back to `uint`

cfergeau commented on cmd/crc/cmd/start.go:
```diff
@@ -70,7 +70,7 @@ func runStart(ctx context.Context) (*types.StartResult, error) {
 		BundlePath:        config.Get(crcConfig.Bundle).AsString(),
 		Memory:            config.Get(crcConfig.Memory).AsUInt(),
-		DiskSize:          config.Get(crcConfig.DiskSize).AsInt(),
+		DiskSize:          config.Get(crcConfig.DiskSize).AsUInt64(),
```
I guess `AsUint` would be enough here?

praveenkumar replied:
switched back to Uint

cfergeau commented on pkg/crc/machine/config/config.go:
```diff
@@ -10,7 +10,7 @@ type MachineConfig struct {
 	Memory            uint
 	CPUs              uint
-	DiskSize          int
+	DiskSize          uint64
```
Can you add a comment specifying the disk size unit? I think it's GiB, in which case `uint` is enough.
You can also add a similar comment to `Memory` in the corresponding commit.

praveenkumar replied:
Done

cfergeau commented on tools/go.mod:
```diff
@@ -4,7 +4,7 @@ go 1.22.1
 require (
 	github.com/cfergeau/gomod2rpmdeps v0.0.0-20210223144124-2042c4850ca8
-	github.com/golangci/golangci-lint v1.59.1
+	github.com/golangci/golangci-lint v1.61.0
```
I would keep this commit until the very end, it's odd to have some commits fixing overflows before this one, and then more commits fixing issues after it.

praveenkumar replied:
Done, push that to end

cfergeau commented on pkg/os/shell/shell_windows.go:
```diff
@@ -43,21 +43,22 @@ func getProcessEntry(pid int) (pe *syscall.ProcessEntry32, err error) {
 
 	if shell == "" {
-		shell, shellppid, err := getNameAndItsPpid(os.Getppid())
+		//#nosec G115 -- pid is coming from os.Getppid() as int
```
Why is this ok to ignore the overflow?
I'd expect some range check before the cast
```
if pid < 0 || pid > math.MaxUint32 {
  // error
}
```

praveenkumar replied:
Do you think `oc.Getppid` return something out of range of int? Shouldn't than it would be checked on os module  level?

cfergeau replied:
If these are the assumptions you want to make to ignore this warning, add them to the `nosec` comment. Imo it does not hurt to have checks to make it very clear.

praveenkumar replied:
Added range check

cfergeau commented on pkg/os/shell/shell_windows_test.go:
```diff
@@ -18,15 +18,18 @@ func TestDetect(t *testing.T) {
 
 func TestGetNameAndItsPpidOfParent(t *testing.T) {
-	shell, _, err := getNameAndItsPpid(os.Getppid())
+	//#nosec G115 -- pid is coming from os.Getppid() as int
```
Same comment as before for these 3 ignores

cfergeau commented on pkg/drivers/vfkit/driver_darwin.go:
```diff
@@ -97,20 +97,21 @@ func (d *Driver) getDiskPath() string {
 
-func (d *Driver) resize(newSize int64) error {
+func (d *Driver) resize(newSize uint64) error {
+	newSizeInt64 := int64(newSize) //#nosec G115 -- os.stat provide the size as int64 and os.Truncate use int64 as arg
```
This explains why there's a warning, but this should be explaining why it's ok to ignore it. This needs range checks before doing the cast.

cfergeau commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -186,7 +186,7 @@ func getTokenForUser(username, password, ip string, ca []byte, clusterConfig *ty
 				}
 				if strings.HasSuffix(hostname, constants.AppsDomain) {
-					port = strconv.Itoa(int(ingressHTTPSPort))
+					port = strconv.FormatUint(uint64(ingressHTTPSPort), 10)
```
Or `fmt.Sprintf("%d", ingressHTTPSPort)`

praveenkumar replied:
yeah, but this is also works so let's go with that.

cfergeau commented on pkg/crc/api/api_client_test.go:
```diff
@@ -195,6 +195,8 @@ func TestConfigGetAll(t *testing.T) {
 		case int:
 			configs[k] = float64(v)
+		case uint:
+			configs[k] = float64(v)
```
Nit, probably could be:
```
case int, uint:
	configs[k] = float64(v)
```

praveenkumar replied:
That doesn't work https://github.com/golang/go/issues/36616

cfergeau commented on pkg/os/shell/shell_windows.go:
```diff
@@ -43,21 +45,26 @@ func getProcessEntry(pid int) (pe *syscall.ProcessEntry32, err error) {
+		if pid < 0 || pid > math.MaxUint32 {
+			return "", fmt.Errorf("integer overflow for pid: %v", pid)
+		}
+		//#nosec G115 -- Above pid is checked for integer overflow and an error is returned
```
With the range check above, the #nosec directive is not needed.

cfergeau commented on pkg/os/shell/shell_windows_test.go:
```diff
@@ -18,15 +19,24 @@ func TestDetect(t *testing.T) {
+	if pid < 0 || pid > math.MaxUint32 {
+		assert.Fail(t, "integer overflow detected")
+	}
+	//#nosec G115 -- Above pid is checked for integer overflow and fail the test
```
I did not test but I'd guess this one is not needed either.

cfergeau replied:
Can be 
```
assert.Positive(t, pid)
assert.Less(t, pid, math.MaxUint32)
```

praveenkumar replied:
> Can be

if we do this then linter again show same error
```
pkg/os/shell/shell_windows_test.go:25:51: G115: integer overflow conversion int -> uint32 (gosec)
	shell, shellppid, err := getNameAndItsPpid(uint32(pid))
	                                                 ^
pkg/os/shell/shell_windows_test.go:27:24: G115: integer overflow conversion int -> uint32 (gosec)
	assert.Equal(t, uint32(pid), shellppid)
	                      ^
pkg/os/shell/shell_windows_test.go:35:43: G115: integer overflow conversion int -> uint32 (gosec)
	shell, _, err := getNameAndItsPpid(uint32(pid))
	                                         ^
```

cfergeau replied:
Ah ok, need to be improved in gosec then.

cfergeau commented on pkg/os/shell/shell_windows_test.go:
```diff
@@ -18,15 +19,24 @@ func TestDetect(t *testing.T) {
+	if pid < 0 || pid > math.MaxUint32 {
+		assert.Fail(t, "integer overflow detected")
+	}
+	//#nosec G115 -- Above pid is checked for integer overflow and fail the test
```
And same here

cfergeau commented on pkg/drivers/vfkit/driver_darwin.go:
```diff
@@ -97,20 +98,25 @@ func (d *Driver) getDiskPath() string {
+	if newSize > math.MaxInt64 {
+		return fmt.Errorf("integer overflow detected for resize: %v", newSize)
+	}
+	//#nosec G115 -- os.stat provide the size as int64 and os.Truncate use int64 as arg. Also range is checked (above)
```
`#nosec G115` is not needed since you added the range check. You can keep the rest of the comment if you think it's useful.

praveenkumar commented:
/hold 

`make check` on windows failing, need to check what's happening.
